booster/library/Booster/Pattern/Rewrite.hs

```diff
@@ -459,6 +417,72 @@ applyRule pat@Pattern{ceilConditions} rule =
+        unclearRequires <-
+            catMaybes <$> mapM (checkConstraint id notAppliedIfBottom pat.constraints) toCheck
+
+        -- unclear conditions may have been simplified and
+        -- could now be syntactically present in the path constraints, filter again
+        stillUnclear <- lift $ filterOutKnownConstraints pat.constraints unclearRequires
+
+        -- check unclear requires-clauses in the context of known constraints (priorKnowledge)
+        solver <- lift $ RewriteT $ (.smtSolver) <$> ask
+        let smtUnclear = do
+                withContext CtxConstraint . withContext CtxAbort . logMessage $
+                    WithJsonMessage (object ["conditions" .= (externaliseTerm . coerce <$> stillUnclear)]) $
+                        renderOneLineText $
+                            "Uncertain about condition(s) in a rule:"
+                                <+> (hsep . punctuate comma . map (pretty' @mods) $ stillUnclear)
+                failRewrite $
+                    RuleConditionUnclear rule . coerce . foldl1 AndTerm $
+                        map coerce stillUnclear
+        SMT.checkPredicates solver pat.constraints mempty (Set.fromList stillUnclear) >>= \case
+            SMT.IsUnknown{} ->
+                smtUnclear -- abort rewrite if a solver result was Unknown
+            SMT.IsInvalid -> do
+                -- requires is actually false given the prior
+                withContext CtxFailure $ logMessage ("Required clauses evaluated to #Bottom." :: Text)
+                RewriteRuleAppT $ pure NotApplied
+            SMT.IsValid ->
+                pure () -- can proceed
+    checkEnsures ::
+        Substitution -> RewriteRuleAppT (RewriteT io) [Predicate]
+    checkEnsures matchingSubst = do
+        -- apply substitution to rule requires
+        let ruleEnsures =
+                concatMap (splitBoolPredicates . coerce . substituteInTerm matchingSubst . coerce) $
+                    Set.toList rule.ensures
+        newConstraints <-
+            catMaybes <$> mapM (checkConstraint id trivialIfBottom pat.constraints) ruleEnsures
+
+        -- check all new constraints together with the known side constraints
+        solver <- lift $ RewriteT $ (.smtSolver) <$> ask
+        (lift $ SMT.checkPredicates solver pat.constraints mempty (Set.fromList newConstraints)) >>= \case
+            SMT.IsInvalid -> do
+                withContext CtxSuccess $ logMessage ("New constraints evaluated to #Bottom." :: Text)
+                RewriteRuleAppT $ pure Trivial
+            _other ->
+                pure ()
+
+        -- if a new constraint is going to be added, the equation cache is invalid
+        unless (null newConstraints) $ do
+            withContextFor Equations . logMessage $
+                ("New path condition ensured, invalidating cache" :: Text)
+
+            lift . RewriteT . lift . modify $ \s -> s{equations = mempty}
+        pure newConstraints
+
```

Comment:
can this be subsumed by the ApplyEquations version?

Reply:
We may find the common denominator, but I'm reluctant to do it as https://github.com/runtimeverification/haskell-backend/pull/4022 will change `checkRequires` in ApplyEquations significantly, and I do not think we want to allow rewrite rules with syntactic `requires` clauses.